src/utils/efd.c

```diff
@@ -44,18 +45,60 @@ int nn_efd_wait (struct nn_efd *self, int timeout)
 {
     int rc;
     struct pollfd pfd;
+    uint64_t expire;
 
-    pfd.fd = nn_efd_getfd (self);
-    pfd.events = POLLIN;
-    if (nn_slow (pfd.fd < 0))
-        return -EBADF;
-    rc = poll (&pfd, 1, timeout);
-    if (nn_slow (rc < 0 && errno == EINTR))
-        return -EINTR;
-    errno_assert (rc >= 0);
-    if (nn_slow (rc == 0))
-        return -ETIMEDOUT;
-    return 0;
+    if (timeout > 0) {
+        expire = nn_clock_ms() + timeout;
+    } else {
+        expire = timeout;
+    }
```

Comment:
I would be happy to make this more explicit -- a timeout of -1 is valid, and results in expire = (uint64_t)-1.  That means an infinite timeout.  This is not a place for an nn_assert().

The unfortunate thing (and it does work this way) is that its not obvious that assigning (int)-1 to a uint64_t results in a value of 0xffffffffffffffffULL (aka (uint64_t)-1).  A more clear way to write this, although slightly less efficient, would be to explicitly check for timeout < 0 and setting expire = (uint64_t)-1.

If you want me to make that change for clarity (at a tiny expense in terms of performance) I'm happy to do so.